pyproject.toml

```diff
@@ -43,6 +43,12 @@ dependencies = [
   "tomli; python_version<'3.11'",
   "userpath!=1.9,>=1.6",
 ]
+optional-dependencies.uv = [
+  "uv>=0.2.27",
+]
+optional-dependencies.virtualenv = [
+  "virtualenv>=20.26.3",
+]
```

Comment:
Looks like it's auto-reformatted by `pyproject-fmt`.

Reply:
Makes me wonder if we should have `optional-dependencies.pip` too (just installs pip, which is no-op most of the time, but explictness is not a bad thing here.)

Reply:
> Makes me wonder if we should have `optional-dependencies.pip` too (just installs pip, which is no-op most of the time, but explictness is not a bad thing here.)

I am not sure whether there will be downsides for it, since it can be quite confusing to users?

Reply:
> > Makes me wonder if we should have optional-dependencies.pip too (just installs pip, which is no-op most of the time, but explictness is not a bad thing here.)

> I am not sure whether there will be downsides for it, since it can be quite confusing to users?

Quite confusing indeed IMHO. I wouldn't add a dependency that never gets used. I would expect only `pip` installed via the specified optional-dependency would be used as the installer. Instead, pip in sharedlibs is the one used.

That being said, the same is true for `uv` & `virtualenv` optional dependencies. In the current state of the PR, the first version present on `PATH` (c.f. `path_to_exec`) gets used. So even if pipx gets installed with those optional dependencies, pipx might fail to use those versions or just fail to find them.
Some precedence should probably be implemented in `path_to_exec` e.g. prefer a version found in the same environment as pipx then find on `PATH`.

If the added complexity in `path_to_exec` is not wanted, then I'd suggest to drop the optional dependencies and to require the user to provide them on `PATH` (which could be documented as `pipx install virtualenv`/`pipx install uv` to bootstrap the tool installation rather than asking for optional dependencies)

Reply:
> If the added complexity in path_to_exec is not wanted, then I'd suggest to drop the optional dependencies and to require the user to provide them on PATH (which could be documented as `pipx install virtualenv`/`pipx install uv` to bootstrap the tool installation rather than asking for optional dependencies)

+1 on this, especially we encourage users to install pipx through their system package managers instead of `pip`, so extra dependencies won't work in this case as well.